FutureFeatures.md

```diff
@@ -66,16 +66,160 @@ Options under consideration:
 
 ## GC/DOM Integration
 
-* Access to certain kinds of Garbage-Collected (GC) objects from variables,
-  arguments, expressions.
-* Ability to GC-allocate certain kinds of GC objects.
-* Initially, things with fixed structure:
-  * JavaScript strings;
-  * JavaScript functions (as callable closures);
-  * Typed Arrays;
-  * [Typed objects](https://github.com/nikomatsakis/typed-objects-explainer/);
-  * DOM objects via WebIDL.
-* Perhaps a rooting API for safe reference from the linear address space.
+To realize the [high-level goals](HighLevelGoals.md) of (1) integrating well
+with the existing web platform and (2) supporting languages other than C++,
+WebAssembly needs to be able to:
+* reference DOM and other Web API objects directly from WebAssembly code;
+* efficiently allocate and manipulate GC objects directly from WebAssembly
+  code; and
+* call Web APIs (passing primitives or DOM/GC/Web API objects) directly from
+  WebAssembly without calling through JS.
+
+These goals can be separated into two complementary sub-features that are
+specified independently of JS and the Web platform but allow a natural
+integration.
+
+### Opaque reference types
+
+The first sub-feature is to extend [module imports](Modules.md#imports-and-exports)
+to allow modules to import *opaque reference types*. "Opaque" means that the
+reference type itself has no structural content and does not, e.g., define any
+methods or fields. Once imported, an opaque reference type can be used in the
+signature of other imported functions. Thus, the point of an opaque reference
+type is to be passed to and returned from exported functions.
+
+Reference types are allowed to be used as the types of locals, parameters
+and return types. Additionally, references would be allowed as operands to
+operators that treat their values as black boxes (`conditional`, `comma`, maybe
+`eq`, etc.). A new `dynamic_cast` operator would be added to allow checked
+casting from any opaque reference type to any other opaque reference type.
+Whether the cast succeeds is up to the host environment; WebAssembly itself
+will define no a priori subtyping relationship.
+
+For reasons of safety and limiting nondeterminism, imported opaque reference
+types would not be able to be loaded from or stored to linear memory where they
+could otherwise be arbitrarily aliased as integers. Instead, a new set of
+operations would be added for allocating, deallocating, loading and storing
+from integer-indexed cells that could hold references and were not aliasable by
+linear memory. There are several important alternatives to consider:
+* LIFO allocation of cells versus (or in addition to) ad hoc explicit
```

Comment:
Are cells more like heap or more like stack slots for GC objects? It's unclear to me whether they're one or both at once. I can imagine use cases where a function just wants to have N reserved GC cells like it has M reserved int32 locals, but also cases where you want to allocate some GC cells and hold on to them indefinitely. Are those both considered cells here, and we just have some lifetime distinction?

I think we need both stack cells and long-lived cells, and they need to have distinct access opcodes. Without that it'll be pretty easy to end up accessing a stack cell after its lifetime has expired. That kinda sucks.

For simplicity's sake, stack cells could be ruled out and they could always be long-lived and manually allocated/deallocated. That's probably the right place to go with the MVP.